### PR TITLE
Always choose the most secure algorithm to check integrity

### PR DESCRIPTION
**Pull request by laeubi**
Currently always all checksums are checked for integrity of a file, this change the behavior by introducing a ranking and only checks the best matching one.

This is another spin-off from #160

### CODE REVIEW

File: bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java [hunk 1]
```diff
@@ -34,13 +34,16 @@ final public class ChecksumVerifier extends MessageDigestProcessingStep {
 	final private String providerName;
 	final private String algorithmId;
 	private final boolean insecureAlgorithm;
+	private int priority;
```
Comment (merks) [inline bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java:37]: Why not final?

File: bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java [hunk 2]
```diff
@@ -34,13 +34,16 @@ final public class ChecksumVerifier extends MessageDigestProcessingStep {
...
 
 	// public to access from tests
-	public ChecksumVerifier(String digestAlgorithm, String providerName, String algorithmId, boolean insecure) {
+	public ChecksumVerifier(String digestAlgorithm, String providerName, String algorithmId, boolean insecure,
+			int priority) {
```
Comment (merks) [inline bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java:41]: If there was any expectation that someone might be using this class directly, leaving the original constructor to call this new one with defaults would be good.  But I doubt there are such...

**Review by merks — APPROVED** [review]
I think the changes are good as is.  I would be good to check that @mickaelistria also agrees is good to just check the most secure available checksum rather than all...

File: bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java (hunk 1 above)
Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java]: Done

File: bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java (hunk 2 above)
Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java:46]: Done.

Comment (merks) [reply bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java:46]: Thanks!

Comment (HannesWell) [reply bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java:46]: Would it be useful to mark that constructor for removal as well?
The new one that is public for tests only could also be marked as deprecated (but not for removal), that that anybody that calls it, without reading the comment, gets it strikedout in the IDE.

Comment (laeubi) [reply bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/processors/checksum/ChecksumVerifier.java:46]: It is internal.. so actually we can delete, modify, .. any time.

**Comment by mickaelistria** [comment]
Yes, checking only the preferred algorithm seems good enough. Most other repository systems (eg rpm, Maven...) do only check 1 integrity algorithm.

**Pull request merged**